File: medium/AddTwoNumbers/cpp/AddTwoNumbers.cpp
#include <iostream>

using std::cout, std::endl;

struct ListNode {
     int val;
     ListNode *next;
     ListNode() : val(0), next(nullptr) {}
     ListNode(int x) : val(x), next(nullptr) {}
     ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode *result = nullptr;
        ListNode *tail = nullptr;
        int carry = 0;

        while (l1 or l2 or carry) {
            int sum = carry;
            if (l1) {
                sum += l1->val;
                l1 = l1->next;
            }
            if (l2) {
                sum += l2->val;
                l2 = l2->next;
            }
            carry = sum >= 10 ? 1 : 0;
            auto temp = new ListNode(sum % 10, nullptr);
            if (!result) {
                result = temp;
                tail = temp;
            } else {
                tail->next = temp;
                tail = tail->next;
            }
        }
        return result;
    }

    void dump(ListNode* l) {   
        cout << "[";
        for (; l ; l = l->next)
            cout << l->val << (l->next ? "," : "]");
    }
};

int main() {
    auto sol = Solution{};
    ListNode *x1 = new ListNode(3);
    ListNode *x2 = new ListNode(4, x1);
    ListNode *x  = new ListNode(2, x2);
    sol.dump(x);
    ListNode *y1 = new ListNode(4);
    ListNode *y2 = new ListNode(6, y1);
    ListNode *y  = new ListNode(5, y2);
    sol.dump(y);

    auto result =  sol.addTwoNumbers(x, y);
    sol.dump(result);

    auto result2 = sol.addTwoNumbers(new ListNode(0), new ListNode(0));
    sol.dump(result2);

    delete x1;
    delete x2;
    delete x;
    delete y1;
    delete y2;
    delete y;
    cout << endl;
}

File: medium/AddTwoNumbers/python/AddTwoNumbers.py
#!/usr/bin/env python3

from logging.config import valid_ident
from typing import List, Optional
from math import log10

# # Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        sum_digits = []
        carry : int = 0
        while l1 is not None or l2 is not None:
            local_sum : int = carry
            if l1 is not None:
                local_sum += l1.val
                l1 = l1.next
            if l2 is not None:
                local_sum += l2.val
                l2 = l2.next
            if local_sum >= 10:
                carry = 1
                local_sum %= 10
            else:
                carry = 0
            sum_digits.append(local_sum)
        if carry == 1:
            sum_digits.append(carry)
        
        sum_digits.reverse()
        result : Optional[ListNode] = None
        for digit in sum_digits:
            result = ListNode(digit, result)
        
        return result
        
    def println(self, l: Optional[ListNode]) -> None:
        node = l
        print("[", end = "")
        while node.next != None:
            print(f"{node.val},",end="")
            node = node.next
        print(f"{node.val}]")

if __name__ == "__main__":
    sol = Solution()
    l1 = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5)))))
    sol.println(l1)
    l2 = ListNode(2, ListNode(8, ListNode(7)))
    sol.println(l2)

    answer = sol.addTwoNumbers(l1, l2)
    sol.println(answer)

    l1 = ListNode(9, ListNode(9, ListNode(9, ListNode(9, ListNode(9, ListNode(9, ListNode(9, ListNode(9, ListNode(9)))))))))
    sol.println(l1)
    l2 = ListNode(9, ListNode(9, ListNode(9, ListNode(9))))
    sol.println(l2)

    answer = sol.addTwoNumbers(l1, l2)
    sol.println(answer)

        
